server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java

```diff
@@ -63,54 +92,98 @@ public void define(WebService.NewController controller) {
-    @Nullable List<String> fields) {
+  private SearchWsResponse buildResponse(List<UserDto> users, Multimap<String, String> groupsByLogin, Map<String, Integer> tokenCountsByLogin,
+    @Nullable List<String> fields, Paging paging) {
+    SearchWsResponse.Builder responseBuilder = newBuilder();
+    users.forEach(user -> responseBuilder.addUsers(towsUser(user, firstNonNull(tokenCountsByLogin.get(user.getLogin()), 0), groupsByLogin.get(user.getLogin()), fields)));
+    responseBuilder.getPagingBuilder()
+      .setPageIndex(paging.pageIndex())
+      .setPageSize(paging.pageSize())
+      .setTotal(paging.total())
+      .build();
+    return responseBuilder.build();
+  }
 
-    json.name("users").beginArray();
-    for (UserDto user : userDtos) {
-      Collection<String> groups = groupsByLogin.get(user.getLogin());
-      userWriter.write(json, user, firstNonNull(tokenCountsByLogin.get(user.getLogin()), 0), groups, fields);
+  private User towsUser(UserDto user, @Nullable Integer tokensCount, Collection<String> groups, @Nullable Collection<String> fields) {
+    User.Builder userBuilder = User.newBuilder()
+      .setLogin(user.getLogin());
+    setIfNeeded(FIELD_NAME, fields, user.getName(), userBuilder::setName);
+    if (userSession.isLoggedIn()) {
+      setIfNeeded(isNeeded(FIELD_AVATAR, fields) && user.getEmail() != null, avatarResolver.create(user), userBuilder::setAvatar);
+      setIfNeeded(FIELD_ACTIVE, fields, user.isActive(), userBuilder::setActive);
+      setIfNeeded(FIELD_LOCAL, fields, user.isLocal(), userBuilder::setLocal);
+      setIfNeeded(FIELD_EXTERNAL_IDENTITY, fields, user.getExternalIdentity(), userBuilder::setExternalIdentity);
+      setIfNeeded(FIELD_EXTERNAL_PROVIDER, fields, user.getExternalIdentityProvider(), userBuilder::setExternalProvider);
+      setIfNeeded(FIELD_TOKENS_COUNT, fields, tokensCount, userBuilder::setTokensCount);
+      setIfNeeded(isNeeded(FIELD_SCM_ACCOUNTS, fields) && !user.getScmAccountsAsList().isEmpty(), user.getScmAccountsAsList(),
+        scm -> userBuilder.setScmAccounts(ScmAccounts.newBuilder().addAllScmAccounts(scm)));
     }
-    json.endArray();
+    if (userSession.isSystemAdministrator()) {
+      setIfNeeded(FIELD_EMAIL, fields, user.getEmail(), userBuilder::setEmail);
+      setIfNeeded(isNeeded(FIELD_GROUPS, fields) && !groups.isEmpty(), groups,
+        g -> userBuilder.setGroups(Groups.newBuilder().addAllGroups(g)));
+    }
+    return userBuilder.build();
   }
 
-  private enum UserDocToLogin implements Function<UserDoc, String> {
-    INSTANCE;
+  private static <PARAM> void setIfNeeded(String field, @Nullable Collection<String> fields, @Nullable PARAM parameter, Function<PARAM, ?> setter) {
+    setIfNeeded(isNeeded(field, fields), parameter, setter);
+  }
 
-    @Override
-    public String apply(@Nonnull UserDoc input) {
-      return input.login();
+  private static <PARAM> void setIfNeeded(boolean condition, @Nullable PARAM parameter, Function<PARAM, ?> setter) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)